File: lib/service/new/chat_service.dart
import 'dart:async';

import 'package:kaonic/data/models/kaonic_new/kaonic_event.dart';
import 'package:kaonic/data/models/kaonic_new/kaonic_event_type.dart';
import 'package:kaonic/data/models/kaonic_new/kaonic_message_event.dart';
import 'package:kaonic/service/new/kaonic_communication_service.dart';
import 'package:rxdart/subjects.dart';

class ChatService {
  ChatService(KaonicCommunicationService kaonicService) {
    _kaonicService = kaonicService;
    _listenMessages();
  }

  late final KaonicCommunicationService _kaonicService;

  /// key is address of chat id
  final _messagesSubject =
      BehaviorSubject<Map<String, List<KaonicEvent>>>.seeded({});

  /// key is contact address,
  /// value is chatUUID
  final _contactChats = <String, String>{};

  String? _address;
  String? _chatId;

  void _listenMessages() {
    _kaonicService.eventsStream
        .where((event) => KaonicEventType.messageEvents.contains(event.type))
        .listen((event) {
      switch (event.type) {
        case KaonicEventType.CHAT_CREATE:
          _putOrUpdateChatId(
            (event.data as MessageEvent).chatId,
            (event.data as MessageEvent).address,
          );
        case KaonicEventType.MESSAGE_TEXT:
        case KaonicEventType.MESSAGE_FILE:
          _handleTextMessage(event);
        // _handleFileMessage(event);
      }
    });
  }

  Stream<List<KaonicEvent>> getChatMessages(String chatId) {
    return _messagesSubject.stream.map((messagesMap) {
      return messagesMap[chatId] ?? [];
    });
  }

  Future<String> createChat(String address) async {
    final chatId = await _kaonicService.createChat(address);

    _contactChats[address] = chatId;
    return chatId;
  }

  void sendTextMessage(String message, String address) async {
    _address = address;
    _kaonicService.sendTextMessage(address, message);
  }

  void sendFileMessage(String filePath, String address) async {
    _address = address;
    final chatId = _contactChats[address];
    if (chatId == null) throw Exception('chatId == null');

    _kaonicService.sendFileMessage(address, filePath, chatId);
  }

  void _handleTextMessage(KaonicEvent message) {
    final data = message.data as MessageEvent;
    final chatId = _address ?? data.address;
    // final chatId = data.chatId;
    _chatId = chatId;
    final currentMap =
        Map<String, List<KaonicEvent>>.from(_messagesSubject.valueOrNull ?? {});
    final messageList = List<KaonicEvent>.from(currentMap[chatId] ?? []);

    final existingMessages = messageList
        .where((msg) =>
            msg.data is MessageEvent &&
            (msg.data as MessageEvent).id == data.id)
        .toList();

    if (existingMessages.isNotEmpty) {
      final index = messageList.indexOf(existingMessages.first);
      if (index != -1) {
        messageList[index] = message;
      }
    } else {
      messageList.add(message);
    }

    currentMap[chatId] = messageList;

    _messagesSubject.add(currentMap);
  }

  void _putOrUpdateChatId(String chatId, String address) {
    _contactChats[address] = chatId;
  }
}

File: android/app/src/main/kotlin/network/beechat/app/kaonic/services/ChatService.kt
package network.beechat.app.kaonic.services

import androidx.compose.runtime.mutableStateMapOf
import com.fasterxml.jackson.databind.ObjectMapper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.launch
import network.beechat.kaonic.models.KaonicEvent
import network.beechat.kaonic.models.KaonicEventType
import network.beechat.kaonic.models.messages.ChatCreateEvent
import network.beechat.kaonic.models.messages.MessageEvent
import network.beechat.kaonic.models.messages.MessageFileEvent
import network.beechat.kaonic.models.messages.MessageTextEvent

class ChatService(scope: CoroutineScope) {
    /**
     * key is address of chat id
     */
    private val messages =
        mutableStateMapOf<String, MutableStateFlow<ArrayList<KaonicEvent<MessageEvent>>>>()

    /**
     * key is contact address,
     * value is chatUUID
     */
    private val contactChats = mutableStateMapOf<String, String>()


    private val objectMapper: ObjectMapper = ObjectMapper()

    init {
        scope.launch {
            KaonicService.events
                .filter { event -> KaonicEventType.messageEvents.contains(event.type) }
                .collect { event ->
                    when (event.data) {
                        is MessageTextEvent -> handleTextMessageEvent(
                            (event.data as MessageTextEvent).chatId,
                            event as KaonicEvent<MessageEvent>
                        )

                        is MessageFileEvent -> handleTextMessageEvent(
                            (event.data as MessageFileEvent).chatId,
                            event as KaonicEvent<MessageEvent>
                        )

                        is ChatCreateEvent ->
                            putOrUpdateChatId(
                                (event.data as ChatCreateEvent).chatId,
                                (event.data as ChatCreateEvent).address,
                            )
                    }
                }
        }
    }

    fun createChatWithAddress(address: String): String {
        if (!contactChats.containsKey(address)) {
            val chatId = java.util.UUID.randomUUID().toString()
            contactChats[address] = chatId
            KaonicService.createChat(address, chatId)
        }

        return contactChats[address]!!
    }

    fun getChatMessages(chatId: String): String {
        val chatMessages = messages.getOrPut(chatId) { MutableStateFlow(arrayListOf()) }
        return  objectMapper.writeValueAsString(chatMessages)
    }

    private fun handleTextMessageEvent(chatId: String, event: KaonicEvent<MessageEvent>) {
        val flow = messages.getOrPut(chatId) { MutableStateFlow(arrayListOf()) }
        val oldList = flow.value
        val existingMessages = oldList.filter {
            it.data is MessageEvent && (it.data as MessageEvent).id == event.data.id
        }
        if (existingMessages.isNotEmpty()) {
            val newList = ArrayList(oldList)
            val index = newList.indexOf(existingMessages.first())
            if (index != -1) {
                newList.removeAt(index)
                newList.add(index, event)
            }
            flow.value = newList
        } else {
            val newList = ArrayList(oldList)
            newList.add(event)
            flow.value = newList
        }
    }

    fun sendTextMessage(message: String, address: String) {
        KaonicService.sendTextMessage(message, address, contactChats[address]!!)
    }

    fun sendFileMessage(filePath: String, address: String) {
        KaonicService.sendFileMessage(filePath, address, contactChats[address]!!)

    }

    private fun putOrUpdateChatId(chatId: String, address: String) {
        contactChats[address] = chatId
    }
}